File: src/renderer/src/lib/posture.ts
import type { DevicePosture, FoldHinge, PostureKind } from '@shared/types'
import { refreshViewport } from './formFactor'
import { createStore } from './store'

export type { DevicePosture, FoldHinge, PostureKind }

/**
 * A foldable's posture as the host reports it (`androidx.window`'s `WindowInfoTracker`, the
 * `FoldingFeature` of the window's layout, `Posture.kt`; OS-11): the device is `flat` (or has
 * no fold in this window), or `halfOpened` with the hinge across the window – a laptop or a
 * book pose. The chrome does not lay itself out by the posture (OS-11: no tabletop layout); the
 * form factor stays the window's width (`formFactor.ts`, re-derived here as the pose changes,
 * since the fold that moves the pose moves the window too). The pose is kept for what reads it
 * – the log, a driver (`data-posture` on the root), a later surface that keeps clear of the
 * hinge – and re-evaluates nothing else by itself.
 */
export const FLAT_POSTURE: DevicePosture = { kind: 'flat', hinge: null }

/**
 * The pose in a word – what the root's `data-posture` says and the log names: `flat`; half
 * opened with a horizontal hinge is `tabletop` (the laptop pose), with a vertical one `book`;
 * half opened with no hinge the host could place is `half-opened`.
 */
export type Pose = 'flat' | 'tabletop' | 'book' | 'half-opened'

export function poseOf(posture: DevicePosture): Pose {
  if (posture.kind !== 'halfOpened') return 'flat'
  if (!posture.hinge) return 'half-opened'
  return posture.hinge.orientation === 'horizontal' ? 'tabletop' : 'book'
}

export interface PostureState {
  posture: DevicePosture
  pose: Pose
}

export const postureStore = createStore<PostureState>(
  { posture: FLAT_POSTURE, pose: 'flat' },
  'posture'
)

/** Two postures name the same pose: the chrome logs and stores a change only when they differ. */
export function samePosture(a: DevicePosture, b: DevicePosture): boolean {
  if (a.kind !== b.kind) return false
  if (a.hinge === null || b.hinge === null) return a.hinge === b.hinge
  return (
    a.hinge.left === b.hinge.left &&
    a.hinge.top === b.hinge.top &&
    a.hinge.right === b.hinge.right &&
    a.hinge.bottom === b.hinge.bottom &&
    a.hinge.orientation === b.hinge.orientation &&
    a.hinge.separating === b.hinge.separating
  )
}

/** The pose and the hinge as one line of the log. */
export function describePosture(posture: DevicePosture): string {
  const pose = poseOf(posture)
  const hinge = posture.hinge
  if (!hinge) return pose
  const size =
    hinge.orientation === 'horizontal'
      ? `${hinge.bottom - hinge.top} tall at y ${hinge.top}`
      : `${hinge.right - hinge.left} wide at x ${hinge.left}`
  return `${pose}, ${hinge.orientation} hinge ${size}${hinge.separating ? ', separating' : ''}`
}

/**
 * The host's `posture` event (and the boot payload's copy, replayed by the bus): the store, the
 * root's `data-posture` and a line in the log when the pose differs from the last one heard – the
 * host reports on a change alone, but a boot replay repeats the last. The viewport re-derives
 * the layout from the window's width in the same breath (`formFactor.ts`): a fold that changed
 * the window has the class follow it, whether or not the resize was heard first. Answers whether
 * anything changed.
 */
export function applyDevicePosture(posture: DevicePosture): boolean {
  const was = postureStore.get().posture
  if (samePosture(was, posture)) return false
  const pose = poseOf(posture)
  postureStore.set({ posture, pose })
  if (typeof document !== 'undefined') document.documentElement.dataset.posture = pose
  console.info(`[zen] posture ${describePosture(posture)}`)
  refreshViewport()
  return true
}

File: android/app/src/main/kotlin/app/zen/chromium/Posture.kt
package app.zen.chromium

import android.app.Activity
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.core.util.Consumer
import androidx.window.java.layout.WindowInfoTrackerCallbackAdapter
import androidx.window.layout.FoldingFeature
import androidx.window.layout.WindowInfoTracker
import androidx.window.layout.WindowLayoutInfo
import org.json.JSONObject

/**
 * A foldable's posture for the chrome (OS-11 / TABLET-08): `androidx.window`'s layout info –
 * the window's [FoldingFeature], when the window lies across a fold – as the `posture` host
 * event, `{ kind: "flat" | "halfOpened", hinge: { left, top, right, bottom, orientation,
 * separating } | null }` with the hinge's bounds in CSS px in the window's coordinates, the way
 * the insets are told ([MainActivity.applyInsets]). The chrome logs the pose and marks its root;
 * it lays itself out by the window's width, not the pose (no tabletop layout), so a fold that
 * moves the pose moves the layout through the resize it brings (`configChanges` keeps the
 * WebViews). Reported at boot ([current], the boot payload's `posture`) and then on a change
 * alone, on the main thread; a device without a fold reports nothing after the boot's `flat`.
 *
 * The report's shape is [Report], free of Android types for JUnit; the tracker's listener is
 * added with the window ([start], `onStart`) and removed as it leaves the screen ([stop]).
 */
class Posture(private val activity: Activity, private val onChange: (JSONObject) -> Unit) {
    private val tracker = WindowInfoTrackerCallbackAdapter(WindowInfoTracker.getOrCreate(activity))
    private var current: JSONObject = Report.flat()
    private var listening = false
    private val listener = Consumer<WindowLayoutInfo> { info -> onLayout(info) }

    /** The posture as last reported (flat until the first layout info arrives). */
    fun current(): JSONObject = current

    /** Listen for the window's layout; the first change after this reaches [onChange]. Idempotent. */
    fun start() {
        if (listening) return
        listening = true
        runCatching {
            tracker.addWindowLayoutInfoListener(activity, ContextCompat.getMainExecutor(activity), listener)
        }.onFailure {
            listening = false
            Log.w(TAG, "posture: window layout info unavailable", it)
        }
    }

    fun stop() {
        if (!listening) return
        listening = false
        runCatching { tracker.removeWindowLayoutInfoListener(listener) }
    }

    private fun onLayout(info: WindowLayoutInfo) {
        val fold = info.displayFeatures.filterIsInstance<FoldingFeature>().firstOrNull()
        val next = if (fold == null) {
            Report.flat()
        } else {
            Report.describe(
                halfOpened = fold.state == FoldingFeature.State.HALF_OPENED,
                horizontal = fold.orientation == FoldingFeature.Orientation.HORIZONTAL,
                left = fold.bounds.left,
                top = fold.bounds.top,
                right = fold.bounds.right,
                bottom = fold.bounds.bottom,
                separating = fold.isSeparating,
                density = activity.resources.displayMetrics.density
            )
        }
        if (Report.same(current, next)) return
        current = next
        Log.i(TAG, "posture ${Report.line(next)}")
        onChange(next)
    }

    /** The `posture` payload: pure, for the JVM tests. */
    object Report {
        const val FLAT = "flat"
        const val HALF_OPENED = "halfOpened"

        /** No fold in this window: a slab, or a foldable whose window lies on one side of the fold. */
        fun flat(): JSONObject = json("kind" to FLAT, "hinge" to null)

        /**
         * A fold across the window: its state as the kind, its bounds (window px) in CSS px by
         * `density`, its orientation and whether it separates the two halves.
         */
        fun describe(
            halfOpened: Boolean,
            horizontal: Boolean,
            left: Int,
            top: Int,
            right: Int,
            bottom: Int,
            separating: Boolean,
            density: Float
        ): JSONObject {
            val d = density.toDouble().takeIf { it > 0 } ?: 1.0
            return json(
                "kind" to if (halfOpened) HALF_OPENED else FLAT,
                "hinge" to json(
                    "left" to left / d,
                    "top" to top / d,
                    "right" to right / d,
                    "bottom" to bottom / d,
                    "orientation" to if (horizontal) "horizontal" else "vertical",
                    "separating" to separating
                )
            )
        }

        /** Two reports name the same pose and hinge: nothing is sent for a repeat. */
        fun same(a: JSONObject, b: JSONObject): Boolean {
            if (a.str("kind") != b.str("kind")) return false
            val ha = a.optJSONObject("hinge")
            val hb = b.optJSONObject("hinge")
            if (ha == null || hb == null) return ha == null && hb == null
            return ha.optDouble("left") == hb.optDouble("left") &&
                ha.optDouble("top") == hb.optDouble("top") &&
                ha.optDouble("right") == hb.optDouble("right") &&
                ha.optDouble("bottom") == hb.optDouble("bottom") &&
                ha.str("orientation") == hb.str("orientation") &&
                ha.optBoolean("separating") == hb.optBoolean("separating")
        }

        /** The report in one line for the log: the pose, then the hinge's orientation and place. */
        fun line(report: JSONObject): String {
            val kind = report.str("kind")
            val hinge = report.optJSONObject("hinge") ?: return kind
            val orientation = hinge.str("orientation")
            val place = if (orientation == "horizontal") {
                "${hinge.optDouble("bottom") - hinge.optDouble("top")} tall at y ${hinge.optDouble("top")}"
            } else {
                "${hinge.optDouble("right") - hinge.optDouble("left")} wide at x ${hinge.optDouble("left")}"
            }
            val separating = if (hinge.optBoolean("separating")) ", separating" else ""
            return "$kind, $orientation hinge $place$separating"
        }
    }

    private companion object {
        const val TAG = "ZenPosture"
    }
}
